### PR TITLE
Send email with office365

### PR DESCRIPTION
**Pull request by michal24509**
add send email to SoftwareUtilities

### CODE REVIEW

File: api/email/send_email.py [hunk 1]
```diff
@@ -14,25 +14,38 @@ class SendEmail:
     def __init__(self, user_name, password):
         self._user_name = user_name
         self._password = password
+        self._connection = self.init_connection()
```
Comment (hChemouni) [inline api/email/send_email.py:17]: It's not necessary

File: api/email/send_email.py [hunk 2]
```diff
@@ -14,25 +14,38 @@ class SendEmail:
...
-        connection.starttls()
-        connection.login(self._user_name, self._password)
         mimemsg = self.create_massage(recipient_address, subject, body)
         time.sleep(2)
+        connection_is_send = True
         try:
+            connection = self.init_connection()
             connection.send_message(mimemsg)
         except Exception as ex:
             Logger().warning(f"send email function get error {ex}")
```
Comment (hChemouni) [inline api/email/send_email.py:44]: Add name of function in warning as SendEmail.send_email have error {ex}

**Review by hChemouni — CHANGES_REQUESTED** [review]

File: api/email/send_email.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import smtplib
+import time
+
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+
+from SoftwareUtilities.logging.logger import Logger
+
+
+class SendEmail:
```
Comment (hChemouni) [inline api/email/send_email.py:10]:
```suggestion
class EmailClient
```

File: api/email/send_email.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import smtplib
+import time
+
```
Comment (hChemouni) [inline api/email/send_email.py:3]: Remove this blank line

File: api/email/send_email.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+class SendEmail:
+    _host = 'smtp.office365.com'
+    _port = 587
+
+    def __init__(self, user_name, password):
+        self._user_name = user_name
+        self._password = password
+
+    def init_connection(self):
```
Comment (hChemouni) [inline api/email/send_email.py:18]: set this function as protected
```suggestion
    def _init_connection(self):
```

File: api/email/send_email.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+from email.mime.text import MIMEText
+
+from SoftwareUtilities.logging.logger import Logger
+
+
+class SendEmail:
+    _host = 'smtp.office365.com'
+    _port = 587
+
+    def __init__(self, user_name, password):
```
Comment (hChemouni) [inline api/email/send_email.py:14]:
```suggestion
    def __init__(self, _user_name, _password):
```

File: api/email/send_email.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        self._user_name = user_name
+        self._password = password
+
+    def init_connection(self):
+        connection_email = smtplib.SMTP(host=self._host, port=self._port)
+        connection_email.starttls()
+        connection_email.login(self._user_name, self._password)
+        return connection_email
+
+    def create_massage(self, recipient_address, subject, body):
```
Comment (hChemouni) [inline api/email/send_email.py:24]:
```suggestion
    def _create_massage(self, _recipient_address, _subject, _body):
```

File: api/email/send_email.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        mime_message = MIMEMultipart()
+        mime_message['From'] = self._user_name
+        mime_message['To'] = recipient_address
+        mime_message['Subject'] = subject
+        mime_message.attach(MIMEText(body, 'plain'))
+        return mime_message
+
+    def send_email(self, recipient_address, subject, body):
+        mimemsg = self.create_massage(recipient_address, subject, body)
```
Comment (hChemouni) [inline api/email/send_email.py:33]:
```suggestion
        mime_message = self.create_massage(_recipient_address=_recipient_address, _subject=_subject, _body=_body)
```

File: api/email/send_email.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    def create_massage(self, recipient_address, subject, body):
+        mime_message = MIMEMultipart()
+        mime_message['From'] = self._user_name
+        mime_message['To'] = recipient_address
+        mime_message['Subject'] = subject
+        mime_message.attach(MIMEText(body, 'plain'))
+        return mime_message
+
+    def send_email(self, recipient_address, subject, body):
```
Comment (hChemouni) [inline api/email/send_email.py:32]:
```suggestion
    def send_email(self, _recipient_address, _subject, _body):
```

File: api/email/send_email.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        mimemsg = self.create_massage(recipient_address, subject, body)
+        time.sleep(2)
+        email_was_successfully_sent = True
+        connection = self.init_connection()
+        try:
+            connection.send_message(mimemsg)
+        except Exception as ex:
+            Logger().warning(f"SendEmail.send_email have error {ex}")
```
Comment (hChemouni) [inline api/email/send_email.py:40]: Update with the new name of class

File: api/email/send_email.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        mime_message = MIMEMultipart()
+        mime_message['From'] = self._user_name
+        mime_message['To'] = recipient_address
+        mime_message['Subject'] = subject
+        mime_message.attach(MIMEText(body, 'plain'))
+        return mime_message
+
+    def send_email(self, recipient_address, subject, body):
+        mimemsg = self.create_massage(recipient_address, subject, body)
+        time.sleep(2)
```
Comment (hChemouni) [inline api/email/send_email.py:34]: Put all code in the try except email_was_successfully_sent.
Change it to be false in default and update to true only in end of try

**Review by hChemouni — CHANGES_REQUESTED** [review]
Change name of to email_client

File: api/email/send_email.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        mime_message.attach(MIMEText(_body, 'plain'))
+        return mime_message
+
+    def send_email(self, _recipient_address, _subject, _body):
+        email_was_successfully_sent = False
+        try:
+            mimemsg = self._create_massage(_recipient_address, _subject, _body)
+            time.sleep(2)
+            connection = self._init_connection()
+            connection.send_message(mimemsg)
```
Comment (hChemouni) [inline api/email/send_email.py:38]: What send email function return ? you need check it and depend of response email_was_successfully_sent to be True or False

**Review by hChemouni — CHANGES_REQUESTED** [review]
Change name of file to be email_client

File: api/email/send_email.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        mime_message.attach(MIMEText(_body, 'plain'))
+        return mime_message
+
+    def send_email(self, _recipient_address, _subject, _body):
+        email_was_successfully_sent = False
+        connection = None
+        try:
+            connection = self._init_connection()
+            mimemsg = self._create_massage(_recipient_address, _subject, _body)
+            time.sleep(2)
```
Comment (hChemouni) [inline api/email/send_email.py:37]: Add explanation why you add the sleep

File: api/email/send_email.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        mime_message['Subject'] = _subject
+        mime_message.attach(MIMEText(_body, 'plain'))
+        return mime_message
+
+    def send_email(self, _recipient_address, _subject, _body):
+        email_was_successfully_sent = False
+        connection = None
+        try:
+            connection = self._init_connection()
+            mimemsg = self._create_massage(_recipient_address, _subject, _body)
```
Comment (hChemouni) [inline api/email/send_email.py:36]:
```suggestion
            mime_message= self._create_massage(_recipient_address, _subject, _body)
```

Comment (hChemouni) [reply api/email/send_email.py:36]: You not change it

File: api/email/send_email.py [hunk 15]
```diff
@@ -34,13 +34,15 @@ def send_email(self, _recipient_address, _subject, _body):
...
+            # Because the amount of sending emails is limited, so there will be a 2-second delay between sending and
+            # sending
             time.sleep(2)
             connection.send_message(mimemsg)
             email_was_successfully_sent = True
         except Exception as ex:
             Logger().warning(f"EmailClient.send_email have error {ex}")
 
-        connection.quit()
+        if connection is not None:
```
Comment (hChemouni) [inline api/email/send_email.py:45]: Why you check it ?

**Review by hChemouni — CHANGES_REQUESTED** [review]

File: api/email/send_email.py (hunk 15 above)
Comment (michal24509) [reply api/email/send_email.py:45]: because if the connection fail the connection is none  and it's passible to quit none

Comment (hChemouni) [reply api/email/send_email.py:45]: it's possible or impossible to quit None ?

Comment (michal24509) [reply api/email/send_email.py:45]: impossible

**Review by hChemouni — APPROVED** [review]

**Pull request merged**